closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -1124,147 +950,864 @@ private void listUsedProfile() {
+                            } else if (existingValue.isBlank()) {
+                                newValue = value;
+                            } else {
+                                newValue = existingValue;
+                            }
+                        }
+                    }
+                } else {
+                    newValue = value;
+                }
+                profileProperties.put(property, newValue);
+                return true;
+            }
+            if (debug != null) {
+                debug.println("Nothing to set. Value of property " + property + " is null.");
+            }
+
+            return false;
+        }
+
+        private String getProperty(String property) {
+            return profileProperties.get(property);
+        }
+
+        private static boolean isPropertyAppendable(String property) {
+            switch (property) {
+                case "jdkCertpathDisabledAlgorithms":
+                case "jdkSecurityLegacyAlgorithms":
+                case "jdkTlsDisabledAlgorithms":
+                case "jdkTlsDisabledNamedCurves":
+                case "jdkTlsLegacyAlgorithms":
+                    return true;
+                default:
+                    return false;
+            }
         }
 
         /**
-         * A class representing the constraints of a provider.
+         * Trim input string if not null.
+         *
+         * @param string the input string
+         * @return the string trimmed or null
          */
-        private static final class Constraint {
-            final String type;
-            final String algorithm;
-            final String attributes;
+        private static String parseProperty(String string) {
+            if (string != null) {
+                string = string.trim();
+            }
 
-            Constraint(String type, String algorithm, String attributes) {
-                super();
-                this.type = type;
-                this.algorithm = algorithm;
-                this.attributes = attributes;
+            return string;
+        }
+
+        private static void checkProviderFormat(String providerInfo, boolean update, boolean isFIPS140_2) {
+            Pattern p = Pattern.compile(
+                "^([A-Za-z0-9\\-\\.]+)"                             // Provider name
+                + "\\s*(\\s[A-Za-z0-9\\$\\{\\}\\./]+)?"             // Configuration file [OPTIONAL]
+                + "\\s*(\\s\\[\\s*"                                 // Constraints [OPTIONAL]
+                    + "([\\+\\-])?\\s*"
+                    + "([A-Za-z0-9\\{\\}\\-\\.\\=\\*\\|:,/_\\s]+)"
+                + "\\])?\\s*$");
+            Matcher m = p.matcher(providerInfo);
+            if (m.find()) {
+                if (!isFIPS140_2) {
+                    String providerName = m.group(1);
+                    if (providerName.indexOf('.') == -1) {
+                        printStackTraceAndExit("Provider can only be specified using"
```

Comment:
Suggest "must" instead of "can only" and make it clear a class name is required:
```java
                        printStackTraceAndExit("Provider must be specified using"
                                + " the fully-qualified class name: " + providerName);
```

Reply:
Changed to suggestion.